src/main/kotlin/com/cognifide/gradle/aem/common/instance/service/pkg/PackageMetadata.kt

```diff
@@ -0,0 +1,75 @@
+package com.cognifide.gradle.aem.common.instance.service.pkg
+
+import com.cognifide.gradle.aem.common.instance.service.repository.Node
+import com.cognifide.gradle.common.utils.Formats
+import java.util.*
+
+@Suppress("TooGenericExceptionCaught")
+class PackageMetadata(val manager: PackageManager, val pkgPath: String) {
+
+    private val logger = manager.logger
+
+    val node by lazy { manager.sync.repository.node("$STORAGE_PATH/${Formats.toHashCodeHex(pkgPath)}") }
+
+    val checksumRemote: String? by lazy { node.takeIf { it.exists }?.properties?.string(CHECKSUM_PROP) }
+
+    fun validChecksum(checksumLocal: String) = (checksumRemote != null) && (checksumLocal == checksumRemote)
+
+    val installedCurrent: Date? by lazy { readInstalledDate() }
+
+    val installedPrevious: Date? get() = node.takeIf { it.exists }?.properties?.date(INSTALLED_PROP)
+
+    val installedExternally: Boolean get() = (installedCurrent != null) && (installedPrevious != null) && (installedCurrent != installedPrevious)
+
+    fun update(checksumLocal: String, action: () -> Unit) = try {
+        saveChecksum(checksumLocal)
+        action()
+        maybeSaveInstalledDate()
+    } catch (e: Exception) {
+        maybeRestoreChecksum()
+        throw e
+    }
+
+    private fun saveChecksum(checksumLocal: String) = save(mapOf(CHECKSUM_PROP to checksumLocal))
+
+    private fun maybeRestoreChecksum() = try {
+        save(mapOf(CHECKSUM_PROP to checksumRemote))
+    } catch (e: Exception) {
+        logger.debug("Cannot restore checksum '$checksumRemote' of package '$pkgPath'!", e)
+    }
+
+    private fun maybeSaveInstalledDate() = try {
```

Comment:
We could use a small comment have why "maybe" - a.k.a. clearly mention that for packages that for example modify OSGi and cause number of bundles to restart - this operational might fail